cosign/verifier.go

```diff
@@ -0,0 +1,334 @@
+		if err != nil {
+			return nil, err
+		}
+		// pkcs11Key, ok := pubKey.(*pkcs11key.Key)
+		// if ok {
+		// 	defer pkcs11Key.Close()
+		// }
+	case vctx.Sk:
+		// TODO: support secure key
+		// sk, err := pivkey.GetKeyWithSlot(vctx.Slot)
+		// defer sk.Close()
+		// if err != nil {
+		// 	return nil, err
+		// }
+		// pubKey, err = sk.Verifier()
+		// if err != nil {
+		// 	return nil, err
+		// }
+	case vctx.CertRef != "":
+		cert, err := s.GetCert(vctx.CertRef)
+		if err != nil {
+			return nil, err
+		}
+		switch {
+		case vctx.CertChain == "" && opts.RootCerts == nil:
+			// If no certChain and no CARoots are passed, the Fulcio root certificate will be used
+			opts.RootCerts, err = fulcioroots.Get()
+			if err != nil {
+				return nil, err
+			}
+			opts.IntermediateCerts, err = fulcioroots.GetIntermediates()
+			if err != nil {
+				return nil, err
+			}
+			pubKey, err = cosign.ValidateAndUnpackCert(cert, opts)
+			if err != nil {
+				return nil, err
+			}
+		case vctx.CertChain != "":
+			// Verify certificate with chain
+			chain, err := s.GetCertChain(vctx.CertChain)
+			if err != nil {
+				return nil, err
+			}
+			pubKey, err = cosign.ValidateAndUnpackCertWithChain(cert, chain, opts)
+			if err != nil {
+				return nil, err
+			}
+		case opts.RootCerts != nil:
+			// Verify certificate with root (and if given, intermediate) certificate
+			pubKey, err = cosign.ValidateAndUnpackCert(cert, opts)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			return nil, err
+		}
+
+		// TODO: fix SCT support
+
+		// if vctx.SCTRef != "" {
+		// 	sct, err := os.ReadFile(filepath.Clean(vctx.SCTRef))
+		// 	if err != nil {
+		// 		return nil, err
+		// 	}
+		// 	opts.IgnoreSCT = vctx.IgnoreSCT
+		// 	opts.SCT = sct
+		// }
+
+	default:
+		// Do nothing. Neither keyRef, c.Sk, nor certRef were set - can happen for example when using Fulcio and TSA.
+		// For an example see the TestAttachWithRFC3161Timestamp test in test/e2e_test.go.
+	}
+	opts.SigVerifier = pubKey
+	opts.MaxWorkers = vctx.MaxWorkers
+
+	return opts, nil
+}
+
+func getgSigandSigDesc(ctx context.Context, opts *ratify.VerifyOptions) (oci.Signature, v1.Hash, error) {
+	signatureDescriptors, err := getSignatureBlobDesc(ctx, opts.Store, opts.Repository, opts.ArtifactDescriptor)
+	if err != nil {
+		return nil, v1.Hash{}, fmt.Errorf("failed to get signature blob descriptor: %w", err)
+	}
+
+	// TODO: check with cosign library for the signature verification update
+	// For now this verifier will only verify the latest signature instead of one of all signatures
+	//
+	// Reference: https://github.com/sigstore/cosign/blob/main/pkg/cosign/verify.go#L722
+	// Reference: https://github.com/sigstore/cosign/blob/main/pkg/cosign/verify.go#L1478
+	numResults := len(signatureDescriptors)
+	if numResults == 0 {
+		return nil, v1.Hash{}, fmt.Errorf("unable to locate reference with artifactType %s", artifactTypeCosign)
+	}
+
+	signatureDesc := signatureDescriptors[numResults-1]
```

Comment:
seems it's an experimental feature for cosign, probably we still need to support the main user scenario. cc: @shizhMSFT